casper/protocols/sharding/sharding_view.py

```diff
@@ -0,0 +1,152 @@
+
+        self.check_forkchoice_atomicity(shards_forkchoice)
+
+        shards_to_build_on = self.select_shards(shards_forkchoice)
+        return {'prev_blocks': {shards_forkchoice[shard_id] for shard_id in shards_to_build_on},
+                'shard_ids': shards_to_build_on}
+
+    def set_child_starting_block(self, tip_block, parent_id, child_id):
+        """Changes the starting block for the forkchoice of a shard"""
+        child_merge_block = self.previous_merge_block_on_shard(
+            tip_block,
+            parent_id,
+            child_id
+        )
+
+        if child_merge_block:
+            self.starting_blocks[child_id] = child_merge_block
+        else:
+            self.starting_blocks[child_id] = self.shard_genesis_blocks[child_id]
+
+    def select_random_shards(self, shards_forkchoice):
+        """Randomly selects a shard to build on, and sometimes selects another child shard"""
+        shards_to_build_on = [r.choice([key for key in self.starting_blocks.keys()])]
+        if (r.randint(0, 1) == 1):
+            child = str(r.randint(0, 1))
+            if shards_to_build_on[0] + child in self.starting_blocks:
+                shards_to_build_on.append(shards_to_build_on[0] + child)
+
+        return set(shards_to_build_on)
+
+    def check_forkchoice_atomicity(self, shards_forkchoice):
+        """Asserts that if a merge block is in the forkchoice for a parent chain
+        then it is in the forkchoice for a child chain"""
+        print("Checking merge block atomicity")
+        for shard_id in sorted(shards_forkchoice):
+            tip = shards_forkchoice[shard_id]
+
+            for child_shard_id in [shard_id + '0', shard_id + '1']:
+                if child_shard_id not in shards_forkchoice:
+                    continue
+
+                merge_block = self.previous_merge_block_on_shard(
+                    tip,
+                    shard_id,
+                    child_shard_id
+                )
+                if merge_block:
+                    assert merge_block.is_in_blockchain(
+                        shards_forkchoice[child_shard_id],
+                        child_shard_id
+                    )
+        print("Passed")
+
+    def previous_merge_block_on_shard(self, starting_block, block_shard_id, merge_shard):
+        """Get the most recent merge block between block_shard_id and merge_shard
+        Starts from starting_block"""
+        assert starting_block.on_shard(block_shard_id)
+        current_block = starting_block
+        while current_block:
+            if current_block.on_shard(merge_shard):
+                assert current_block.is_merge_block
+                return current_block
+            current_block = current_block.prev_block(block_shard_id)
+        return None
+
+    def update_safe_estimates(self, validator_set):
+        """Checks safety on messages in views forkchoice, and updates last_finalized_block"""
+        # check the safety of the top shard!
+        pass
```

Comment:
What's going on here?

Reply:
ah, does oracle not work?

Reply:
Yeah. Doesn't work yet. Have to do a bit of thinking on how to fix this up (don't think it should be bad).